src/tracer.js

```diff
@@ -0,0 +1,218 @@
+        this._serviceName = serviceName;
+        this._reporter = reporter || new NoopReporter();
+        this._sampler = sampler || new NoopSampler();
+        this._host = Utils.createEndpoint(serviceName, myLocalIp(), port);
+        this._injectors = {};
+        this._extractors = {};
+
+        let textCodec = new TextMapCodec(false);
+        this.registerInjector(opentracing.FORMAT_TEXT_MAP, textCodec);
+        this.registerExtractor(opentracing.FORMAT_TEXT_MAP, textCodec);
+
+        let httpCodec = new TextMapCodec(true);
+        this.registerInjector(opentracing.FORMAT_HTTP_HEADERS, httpCodec);
+        this.registerExtractor(opentracing.FORMAT_HTTP_HEADERS, httpCodec);
+    }
+
+    _createSpan(operationName: string,
+                spanContext: SpanContext,
+                isClient: boolean,
+                peer: Endpoint): Span {
+
+        let span = new Span(
+            this,
+            operationName,
+            spanContext,
+            Utils.getTimestampMicros(),
+            isClient,
+            peer
+        );
+
+        span.log({
+            'event': isClient ? thrift.CLIENT_SEND : thrift.SERVER_RECV,
+            'timestamp': span._timestamp
+        });
+
+        return span;
+    }
+
+    _report(span: Span): void {
+        span.addTags(this._tracerTags);
+        this._reporter.report(span);
+    }
+
+    registerInjector(format: string, injector: Injector): void {
+        this._injectors[format] = injector;
+    }
+
+    registerExtractor(format: string, extractor: Extractor): void {
+        this._extractors[format] = extractor;
+    }
+
+    /**
+    * The method for creating a root or child span.
+    *
+    * @param {object} fields - the fields to set on the newly created span.
+    * @param {string} fields.operationName - the name to use for the newly
+    *        created span. Required if called with a single argument.
+    * @param {SpanContext} [fields.childOf] - a parent SpanContext (or Span,
+    *        for convenience) that the newly-started span will be the child of
+    *        (per REFERENCE_CHILD_OF). If specified, `fields.references` must
+    *        be unspecified.
+    * @param {object} [fields.tags] - set of key-value pairs which will be set
+    *        as tags on the newly created Span. Ownership of the object is
+    *        passed to the created span for efficiency reasons (the caller
+    *        should not modify this object after calling startSpan).
+    * @param {number} [fields.startTime] - a manually specified start time for
+    *        the created Span object. The time should be specified in
+    *        milliseconds as Unix timestamp. Decimal value are supported
+    *        to represent time values with sub-millisecond accuracy.
+    * @return {Span} - a new Span object.
+    **/
+    startSpan(fields: spanFields): Span {
+        let randomId = Utils.getRandom64();
+        let isClient = fields.isClient || !!fields.childOf || false;
```

Comment:
fields.isClient is not a thing. ChildOf has no bearing on isClient. Compare with go logic